ops/main.py

```diff
@@ -369,97 +337,282 @@ def _should_use_controller_storage(db_path: Path, meta: CharmMeta) -> bool:
+        charm_type = self.charm_spec.charm_type
+
+        charm = charm_type(framework)
+        dispatcher.ensure_event_links(charm)
+        return charm
+
+    @staticmethod
+    def _setup_model_backend():
+        # this is a standalone method to facilitate overriding by subclassing in testing
+        return ops.model._ModelBackend()
+
+    @staticmethod
+    def _setup_root_logging(backend: ops.model._ModelBackend):
+        debug = "JUJU_DEBUG" in os.environ
+        setup_root_logging(backend, debug=debug)
+
+        # our hello world
+        logger.debug("ops %s up and running.", ops.__version__)  # type:ignore
+
+    def _setup_storage(self, dispatcher: _Dispatcher):
+        charm_state_path = self.charm_spec.charm_root / self.CHARM_STATE_FILE
+
+        use_juju_for_storage = self._use_juju_for_storage
+        if use_juju_for_storage and not ops.storage.juju_backend_available():
+            # raise an exception; the charm is broken and needs fixing.
+            msg = 'charm set use_juju_for_storage=True, but Juju version {} does not support it'
+            raise RuntimeError(msg.format(JujuVersion.from_environ()))
+
+        if use_juju_for_storage is None:
+            use_juju_for_storage = _should_use_controller_storage(
+                charm_state_path,
+                self._charm_meta
+            )
+        elif use_juju_for_storage:
+            warnings.warn("Controller storage is deprecated; it's intended for "
+                          "podspec charms and will be removed in a future release.",
+                          category=DeprecationWarning)
+
+        if use_juju_for_storage and dispatcher.is_restricted_context():
             # TODO: jam 2020-06-30 This unconditionally avoids running a collect metrics event
             #  Though we eventually expect that Juju will run collect-metrics in a
-            #  non-restricted context. Once we can determine that we are running collect-metrics
-            #  in a non-restricted context, we should fire the event as normal.
+            #  non-restricted context. Once we can determine that we are running
+            #  collect-metrics in a non-restricted context, we should fire the event as normal.
             logger.debug('"%s" is not supported when using Juju for storage\n'
                          'see: https://github.com/canonical/operator/issues/348',
                          dispatcher.event_name)
             # Note that we don't exit nonzero, because that would cause Juju to rerun the hook
-            return
-        store = ops.storage.JujuStorage()
-    else:
-        store = ops.storage.SQLiteStorage(charm_state_path)
-    framework = ops.framework.Framework(store, charm_dir, meta, model,
-                                        event_name=dispatcher.event_name)
-    framework.set_breakpointhook()
-    try:
-        charm = charm_class(framework)
-        dispatcher.ensure_event_links(charm)
+            exit(0)
```

Comment:
Seems very strange to call `exit()` in a method. Is there a better way to do this? Return a value, or raise an exception?